### PR TITLE
Post Featured Image: New design for Replace and Remove buttons

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
Updates the post editor's Featured Image panel to use ~the dropdown-based design that was built in https://github.com/WordPress/gutenberg/pull/42352 but never shipped~ a new design.

## Why?
- The current approach of having _Replace Image_ and _Remove featured image_ buttons below the image looks disoriented.
- ~It gives us room to add an _Upload file_ button.~

## Testing Instructions
1. Create or edit a post.
2. Open the settings sidebar.
3. Add a featured image.

## Screenshots or screencast 

| Before | After |
| - | - |
| Screenshot 2023-05-03 at 11 57 06 | Screenshot 2023-05-03 at 11 57 30 |

https://user-images.githubusercontent.com/612155/235820264-f6819b0f-71a8-4859-9d61-33650e6629e8.mp4

### CODE REVIEW

File: packages/editor/src/components/post-featured-image/index.js [hunk 1]
```diff
@@ -160,83 +170,111 @@ function PostFeaturedImage( {
...
+											className="editor-post-featured-image__dropdown"
+											popoverProps={ {
+												placement: 'left-start',
+												shift: true,
+											} }
+											renderToggle={ ( {
+												isOpen,
+												onToggle,
+											} ) => (
+												<Button
```
Comment (noisysocks) [inline packages/editor/src/components/post-featured-image/index.js]: The amount of indentation here is getting crazy 😅 Will clean it up if we decide to proceed with the design.

File: packages/editor/src/components/post-featured-image/index.js [hunk 2]
```diff
@@ -160,83 +170,111 @@ function PostFeaturedImage( {
...
-										variant="secondary"
-									>
-										{ __( 'Replace Image' ) }
-									</Button>
+								render={ ( { openFileDialog } ) => (
+									<div className="editor-post-featured-image__container">
+										<Dropdown
+											className="editor-post-featured-image__dropdown"
+											popoverProps={ {
+												placement: 'left-start',
```
Comment (noisysocks) [inline packages/editor/src/components/post-featured-image/index.js]: This doesn't look the greatest on mobile. Maybe should consider `'overlay'` or `'bottom'` when the viewport is small.

**Comment by aurooba** [comment]
A thought: what if when you click/select, it transforms into the placeholder with ability to upload, select from the media library, and remove? 

This is not a totally accurate mockup, just a quick one I did in the browswer, but I think it gets the idea acros.
a featured image component resembling the placeholder component for the image block

(I'm sure there's a conversation around this elsewhere that I haven't seen so please feel free to dismiss. I'd appreciate a link to context if there is some!)

File: packages/editor/src/components/post-featured-image/index.js (hunk 2 above)
Comment (ramonjd) [reply packages/editor/src/components/post-featured-image/index.js]: Here's what I'm seeing:

Screenshot 2023-05-03 at 4 24 31 pm

Would it be acceptable to hide the popover in narrower viewports and have a modified version of what's on trunk?

Screenshot 2023-05-03 at 4 29 46 pm

Comment (ramonjd) [reply packages/editor/src/components/post-featured-image/index.js]: > Would it be acceptable to hide the popover in narrower viewports and have a modified version of what's on trunk?

Actually, now that I think of it, the replace image button is redundant since the image is a button and triggers the same action.

The remove link could just be a ❌  icon too.

🤷 What do I know.

**Comment by jasmussen** [comment]
Nice, thanks for doing that. This is important polish. I feel like it's nearly there. Before:

before

After:

featured

Overall this is much better than before, with a better footprint. A few things, though. In this state, we don't need the flyout. You can drag and drop images, and you can upload inside the media library modal anyway. We can remove the upload and go straight to the media library:

Screenshot 2023-05-03 at 09 09 47

In fact, if we remove the upload button, we are left with just two actions, open media library and remove. So can we remove the dropdown entirely, and have two buttons? Just an edit button, and a trash button, as overlay buttons in the top right corner. Clicking the thumbnail just opens the media library:

Screenshot 2023-05-03 at 09 03 53

As an overlay button, it's important that it overlays the image, it can't just sit in the white space when the image doesn't fill the area, as it does in square or portrait orientations:

Screenshot 2023-05-03 at 09 06 33

Additionally, portrait featured images get very tall, there's no reason to. I meant to go in and push a fix, so instead of these:

Screenshot 2023-05-03 at 09 16 21

Screenshot 2023-05-03 at 09 18 16

we get this:

Screenshot 2023-05-03 at 09 20 34

Screenshot 2023-05-03 at 09 25 41

I pushed a small fix to the focus style and radius as well.

Note that there is a bottom margin we need to remove:

Screenshot 2023-05-03 at 09 25 49

It appears here:
Screenshot 2023-05-03 at 09 25 53

But I can't find it in the codebase. Can you find it and remove it? We might also not need the position relative.

So to summarize:

* This is nearly ready
* Let's add a trash button next to the edit button, and have both the edit button and the thumbnail itself open the media library directly. No need for a dropdown.

Thank you Robert!

File: packages/editor/src/components/post-featured-image/menu.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+export default function PostFeaturedImageMenu( {
+	removeImageLabel,
+	onClose,
+	onOpenMediaLibrary,
+	onOpenFileDialog,
+	onRemoveImage,
+} ) {
+	return (
+		<NavigableMenu className="editor-post-featured-image__menu">
```
Comment (ntsekouras) [inline packages/editor/src/components/post-featured-image/menu.js]: Should we use `MediaReplaceFlow` here? The icons are also at the right there and here are on the left.

**Comment by aurooba** [comment]
> can we remove the dropdown entirely, and have two buttons? Just an edit button, and a trash button, as overlay buttons in the top right corner. Clicking the thumbnail just opens the media library.

Popping in as a user to say that I would very much prefer to see ^ this rather than the flyout menu. Some kind of overlay for editing and deleting is a better user experience. Flyout menus that live _outside_ the Inspector visually but are triggered from the Inspector can become problematic very quickly.

File: packages/editor/src/components/post-featured-image/index.js (hunk 2 above)
Comment (mtias) [reply packages/editor/src/components/post-featured-image/index.js]: For narrower viewports maybe we could do something like this (or with the two buttons below the image, but looking the same)

image

File: packages/editor/src/components/post-featured-image/index.js [hunk 4]
```diff
@@ -160,83 +170,111 @@ function PostFeaturedImage( {
...
+															?.set_featured_image ||
+															DEFAULT_SET_FEATURE_IMAGE_LABEL ) }
+												</Button>
+											) }
+											renderContent={ ( { onClose } ) => (
+												<PostFeaturedImageMenu
+													removeImageLabel={
+														postType?.labels
+															?.remove_featured_image ||
+														DEFAULT_REMOVE_FEATURE_IMAGE_LABEL
```
Comment (mtias) [inline packages/editor/src/components/post-featured-image/index.js]: Can this be just "Remove" instead of repeating featured image again?

File: packages/editor/src/components/post-featured-image/menu.js (hunk 3 above)
Comment (noisysocks) [reply packages/editor/src/components/post-featured-image/menu.js]: Not using a dropdown anymore!

**Comment by noisysocks** [comment]
Thanks for the feedback @jasmussen. I think I addressed all of it. Let me know if not.

https://user-images.githubusercontent.com/612155/236125417-8126c216-137d-42e4-88df-797386b9598d.mp4

Not sure if you meant for the aspect ratio to be like this on narrow viewports?

Screenshot 2023-05-04 at 16 14 23

**Comment by draganescu** [comment]
I'm glad we got rid of the flyout 🎉

**Comment by aurooba** [comment]
If the icons were centred on the bottom, then they would appear on top of the image even if the selected image is a square and doesn't take up the entire featured image div., unless I'm misinterpreting and the plan is to make all images fit the entire featured image div. 🤔

**Comment by jasmussen** [comment]
Looks good to me, thanks for the followup!

This looks pretty much ready to go to me, though I think I saw Matías mention Replace/Remove buttons instead of icons? I don't personally have a strong opinion there and can go either way. 

Tabbing works too:

tabbing

That said, notice at the end of the GIF, there appears to be a focus loss if you invoke either of the two Replace/Remove buttons. I.e. if you tab to the thumbnail itself, press enter, you open the media library modal. Escape closes it again, and sets focus back on the thumbnail. However if you tab to the Edit button, press Enter, then Escape to close the modal, the focus is on `body`, same with the trash button. Can you restore focus back to the thumbnail?

**Comment by noisysocks** [comment]
> This looks pretty much ready to go to me, though I think I saw Matías mention Replace/Remove buttons instead of icons? I don't personally have a strong opinion there and can go either way.

The design he shared in https://github.com/WordPress/gutenberg/pull/50269#discussion_r1184150798? I've gone ahead and implemented it in 258b603. I also added a transition in 4706919 as that felt right. Feel free to tweak the glass effect, transition, or revert the commits altogether if you prefer the icon design.

https://user-images.githubusercontent.com/612155/236355747-3fb8ae48-a375-4b97-8c37-962145dad3cf.mp4

> However if you tab to the Edit button, press Enter, then Escape to close the modal, the focus is on `body`, same with the trash button. Can you restore focus back to the thumbnail?

Fixed in 3909b6e!

**Review by jasmussen — APPROVED** [review]
This tests well for me, and is a substantial improvement over trunk. Let's get it in.

**Comment by mtias** [comment]
Let's get it in and iterate, it's nice to see all the QoL improvements to the post settings panel!
